### PR TITLE
Oss 865 Implement new driver exceptions hierarchy

### PR DESCRIPTION
**Pull request by shestakovg**
### Overview
 Implemented new driver exceptions hierarchy
### Jira ticket
https://faunadb.atlassian.net/browse/OSS-865

### CODE REVIEW

File: FaunaDB.Client/Errors/FaunaException.cs [hunk 1]
```diff
@@ -10,114 +8,158 @@ namespace FaunaDB.Errors
...
+    {
+        public InvalidScope(int httpStatusCode, string message, string[] position) : base(httpStatusCode, ExceptionCodes.InvalidScope, message, position) { }
+    }
+
+    public class InvalidToken : FaunaException
+    {
+        public InvalidToken(int httpStatusCode, string message, string[] position) : base(httpStatusCode, ExceptionCodes.InvalidToken, message, position) { }
+    }
+
+    public class CallError : FaunaException
```
Comment (parkhomenko) [inline FaunaDB.Client/Errors/FaunaException.cs:119]: To be consistent with other drivers, let's call it a `FunctionCallException`

File: FaunaDB.Client/Errors/FaunaException.cs [hunk 2]
```diff
@@ -10,114 +8,158 @@ namespace FaunaDB.Errors
...
+    {
+        public CallError(int httpStatusCode, string message, string[] position) : base(httpStatusCode, ExceptionCodes.CallError, message, position) { }
```
Comment (parkhomenko) [inline FaunaDB.Client/Errors/FaunaException.cs:121]: There is also an internal `cause` object that needs to be added here, check out a Java driver for this:https://github.com/fauna/faunadb-jvm/blob/OSS-864/faunadb-java/src/main/java/com/faunadb/client/FaunaClient.java#L536

File: FaunaDB.Client/Errors/FaunaException.cs [hunk 3]
```diff
@@ -10,114 +8,158 @@ namespace FaunaDB.Errors
...
+    }
+
+    public class InstanceAlreadyExists : FaunaException
+    {
+        public InstanceAlreadyExists(int httpStatusCode, string message, string[] position) : base(httpStatusCode, ExceptionCodes.InvaliRef, message, position) { }
+    }
+
+    public class ValidationFailed : FaunaException
+    {
+        public ValidationFailed(int httpStatusCode, string message, string[] position) : base(httpStatusCode, ExceptionCodes.ValidationFailed, message, position) { }
```
Comment (parkhomenko) [inline FaunaDB.Client/Errors/FaunaException.cs:61]: There is an internal `failure` object that needs to be added here:
https://github.com/fauna/faunadb-jvm/blob/OSS-864/faunadb-java/src/main/java/com/faunadb/client/FaunaClient.java#L570

File: FaunaDB.Client/Errors/FaunaException.cs [hunk 4]
```diff
@@ -10,114 +8,158 @@ namespace FaunaDB.Errors
...
-            : base(response) { }
+        public SchemaNotFound(int httpStatusCode, string message, string[] position) : base(httpStatusCode, ExceptionCodes.InvaliRef, message, position) { }
+    }
+
+    public class InstanceAlreadyExists : FaunaException
+    {
+        public InstanceAlreadyExists(int httpStatusCode, string message, string[] position) : base(httpStatusCode, ExceptionCodes.InvaliRef, message, position) { }
+    }
+
+    public class ValidationFailed : FaunaException
```
Comment (parkhomenko) [inline FaunaDB.Client/Errors/FaunaException.cs:59]: Let's be consistent with other drivers and add an `Exception` suffix to each class here, so this one, for example, will be named `ValidationFailedException`

**Review by parkhomenko — CHANGES_REQUESTED** [review]
Everything looks great!
I've left some comments. Also, check out the doc for exceptions you missed to test:
https://docs.google.com/spreadsheets/d/1j-ee0Ipp8MzSsXHiP7Z-gs0hv4-ejm6qfI-uETVfsvA/edit#gid=0

File: FaunaDB.Client/Errors/ExceptionResolver.cs [hunk 5]
```diff
@@ -0,0 +1,101 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace FaunaDB.Errors
+{
+    public class ExceptionResolver
+    {
+        public static FaunaException[] Resolve(QueryErrorResponse responce, int httpStatusCode)
```
Comment (freels) [inline FaunaDB.Client/Errors/ExceptionResolver.cs:9]:
```suggestion
        public static FaunaException[] Resolve(QueryErrorResponse response, int httpStatusCode)
```

File: FaunaDB.Client/Errors/ExceptionResolver.cs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+            List<FaunaException> exceptions = new List<FaunaException>();
+            foreach (var error in responce.Errors)
+            {
+                string[] exceptionPositions = new string[error.Position.Count];
+                string[] responsePosition = error.Position.ToArray();
+                Array.Copy(responsePosition, exceptionPositions, responsePosition.Length);
+
+                switch (error.Code)
+                {
+                    case ExceptionCodes.InvaliRef:
```
Comment (freels) [inline FaunaDB.Client/Errors/ExceptionResolver.cs:20]:
```suggestion
                    case ExceptionCodes.InvalidRef:
```

File: FaunaDB.Client/Errors/ExceptionResolver.cs [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+namespace FaunaDB.Errors
+{
+    public class ExceptionResolver
+    {
+        public static FaunaException[] Resolve(QueryErrorResponse responce, int httpStatusCode)
+        {
+            List<FaunaException> exceptions = new List<FaunaException>();
+            foreach (var error in responce.Errors)
+            {
+                string[] exceptionPositions = new string[error.Position.Count];
```
Comment (freels) [inline FaunaDB.Client/Errors/ExceptionResolver.cs:14]: singular?
```suggestion
                string[] exceptionPosition = new string[error.Position.Count];
```

File: FaunaDB.Client/Errors/ExceptionCode.cs [hunk 8]
```diff
@@ -0,0 +1,33 @@
+﻿using System.ComponentModel;
+
+namespace FaunaDB.Errors
+{
+    public class ExceptionCodes
+    {
+        public const string InvalidExpression = "invalid expression";
+        public const string InvalidUrlParameter = "invalid url parameter";
+        public const string SchemaNotFound = "schema not found";
```
Comment (freels) [inline FaunaDB.Client/Errors/ExceptionCode.cs:9]: remove (and related exception class)

File: FaunaDB.Client/Errors/ExceptionCode.cs [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        public const string TransactionAborted = "transaction aborted";
+        public const string InvalidWriteTime = "invalid write time";
+        public const string InvalidArgument = "invalid argument";
+        public const string InvaliRef = "invalid ref";
+        public const string MissingIdentity = "missing identity";
+        public const string InvalidScope = "invalid scope";
```
Comment (freels) [inline FaunaDB.Client/Errors/ExceptionCode.cs:15]: remove (and related exception class)

File: FaunaDB.Client/Errors/ExceptionCode.cs [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+        public const string CallError = "call error";
+        public const string StackOverflow = "stack overflow";
+        public const string PermissionDenied = "permission denied";
+        public const string AuthenticationFailed = "authentication failed";
+        public const string ValueNotFound = "value not found";
+        public const string InstanceNotFound = "instance not found";
+        public const string InstanceAlreadyExists = "instance already exists";
+        public const string ValidationFailed = "validation failed";
+        public const string InstanceNotUnique = "instance not unique";
+        public const string InvalidObjectInContainer = "invalid object in container";
```
Comment (freels) [inline FaunaDB.Client/Errors/ExceptionCode.cs:26]: remove (and related exception class)

File: FaunaDB.Client/Errors/ExceptionCode.cs [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+        public const string MoveDatabaseError = "move database error";
```
Comment (freels) [inline FaunaDB.Client/Errors/ExceptionCode.cs:27]: remove (and related exception class)

File: FaunaDB.Client/Errors/ExceptionCode.cs [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+        public const string RecoveryFailed = "recovery failed";
```
Comment (freels) [inline FaunaDB.Client/Errors/ExceptionCode.cs:28]: remove (and related exception class)

File: FaunaDB.Client/Client/FaunaClient.cs [hunk 13]
```diff
@@ -184,23 +184,7 @@ internal static void RaiseForStatusCode(RequestResult resultRequest)
...
-                case 404:
-                    throw new NotFound(response);
-                case 500:
-                    throw new InternalError(response);
-                case 503:
-                    throw new UnavailableError(response);
-                default:
-                    throw new UnknowException(response);
-            }
+            throw ExceptionResolver.Resolve(response, statusCode).FirstOrDefault();
```
Comment (freels) [inline FaunaDB.Client/Client/FaunaClient.cs:187]: did we ever handle the case where the response body was not valid JSON? We need to handle that here, throwing the correct exception based on the HTTP status code as a fallback. See the Java driver for an example.

**Review by freels — CHANGES_REQUESTED** [review]

**Review by parkhomenko — APPROVED** [review]

**Pull request merged**